File: ai-engine/utils/trust_score.py
"""
Trust Score & Confidence Level Utilities — EDDS AI Engine

Computes trust_score, temporal consistency, and self-aware confidence
level from existing detection outputs. Added as a non-intrusive module
that is called at the end of detection without modifying core logic.
"""
import numpy as np
from typing import List, Optional, Tuple


def compute_agreement_score(fake_probabilities: List[float]) -> float:
    """
    Compute model agreement score (0–1) from individual model fake_probabilities.
    Higher score = models agree. Lower score = models disagree.

    Uses variance-based approach when multiple models exist.
    """
    if len(fake_probabilities) < 2:
        # Single model or no models — default moderate agreement
        return 0.80

    variance = float(np.var(fake_probabilities))
    # Max realistic variance for probabilities in [0,1] is ~0.25
    # Map: variance=0 → agreement=1.0, variance>=0.25 → agreement=0.0
    agreement = max(0.0, 1.0 - (variance / 0.25))
    return round(agreement, 4)


def compute_trust_score(
    confidence: float,
    agreement_score: float,
    forensic_score: Optional[float] = None,
) -> float:
    """
    Composite trust score (0–100) combining:
      - 50% detection confidence
      - 30% model agreement
      - 20% forensic consistency

    Formula:
      trust_score = (0.5 * confidence + 0.3 * agreement + 0.2 * forensic) * 100
    """
    # If no forensic score is available, use a reasonable placeholder
    if forensic_score is None:
        forensic_score = 0.75  # neutral default

    raw = (0.5 * confidence) + (0.3 * agreement_score) + (0.2 * forensic_score)
    return round(min(100.0, max(0.0, raw * 100)), 2)


def compute_temporal_variance(frame_scores: List[float]) -> Tuple[Optional[float], Optional[str]]:
    """
    Compute temporal consistency metrics for video analysis.

    Returns:
        (temporal_variance, temporal_label)
        Both are None for images (no frames).
    """
    if not frame_scores or len(frame_scores) < 2:
        return None, None

    variance = float(np.var(frame_scores))
    variance = round(variance, 4)

    if variance < 0.05:
        label = "Stable"
    elif variance <= 0.15:
        label = "Moderate Variation"
    else:
        label = "High Instability"

    return variance, label


def compute_confidence_level(
    confidence_pct: float,
    temporal_variance: Optional[float] = None,
) -> str:
    """
    Self-aware confidence output.

    Args:
        confidence_pct: Overall confidence as a percentage (0–100).
        temporal_variance: Variance across frames (None for images).

    Returns:
        Human-readable confidence label.
    """
    high_variance = (temporal_variance is not None and temporal_variance > 0.15)

    if confidence_pct < 50 or high_variance:
        return "Low Confidence - Human Review Recommended"
    elif confidence_pct < 75:
        return "Moderate Confidence"
    else:
        # Also downgrade if moderate temporal variance
        if temporal_variance is not None and temporal_variance > 0.05:
            return "Moderate Confidence"
        return "High Confidence"

File: backend/src/utils/trustScore.js
/**
 * Trust Score Utilities — EDDS Backend
 *
 * Mirrors the Python ai-engine/utils/trust_score.py logic.
 * Used to enrich the backend's simulation-mode results with
 * trust_score, temporal_variance, temporal_label, and confidence_level.
 */

/**
 * Compute model agreement score (0–1).
 * Higher = models agree more.
 */
function computeAgreementScore(fakeProbabilities) {
    if (!fakeProbabilities || fakeProbabilities.length < 2) {
        return 0.80; // single-model default
    }
    const mean = fakeProbabilities.reduce((a, b) => a + b, 0) / fakeProbabilities.length;
    const variance = fakeProbabilities.reduce((sum, p) => sum + (p - mean) ** 2, 0) / fakeProbabilities.length;
    return Math.max(0, Math.round((1 - variance / 0.25) * 10000) / 10000);
}

/**
 * Composite trust score (0–100).
 * Formula: (0.5 * confidence + 0.3 * agreement + 0.2 * forensic) * 100
 */
function computeTrustScore(confidence, agreementScore, forensicScore = 0.75) {
    const raw = (0.5 * confidence) + (0.3 * agreementScore) + (0.2 * forensicScore);
    return Math.round(Math.min(100, Math.max(0, raw * 100)) * 100) / 100;
}

/**
 * Temporal variance & label (video only).
 * Returns { variance, label } or { variance: null, label: null } for images.
 */
function computeTemporalVariance(frameScores) {
    if (!frameScores || frameScores.length < 2) {
        return { variance: null, label: null };
    }
    const mean = frameScores.reduce((a, b) => a + b, 0) / frameScores.length;
    const variance = Math.round(
        (frameScores.reduce((sum, s) => sum + (s - mean) ** 2, 0) / frameScores.length) * 10000
    ) / 10000;

    let label;
    if (variance < 0.05) label = 'Stable';
    else if (variance <= 0.15) label = 'Moderate Variation';
    else label = 'High Instability';

    return { variance, label };
}

/**
 * Self-aware confidence label.
 */
function computeConfidenceLevel(confidencePct, temporalVariance = null) {
    const highVariance = temporalVariance !== null && temporalVariance > 0.15;

    if (confidencePct < 50 || highVariance) {
        return 'Low Confidence - Human Review Recommended';
    }
    if (confidencePct < 75) {
        return 'Moderate Confidence';
    }
    if (temporalVariance !== null && temporalVariance > 0.05) {
        return 'Moderate Confidence';
    }
    return 'High Confidence';
}

module.exports = {
    computeAgreementScore,
    computeTrustScore,
    computeTemporalVariance,
    computeConfidenceLevel,
};
